pkg/datapath/sockets/sockets.go

```diff
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package sockets
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/sys/unix"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	sizeofSocketID      = 0x30
+	sizeofSocketRequest = sizeofSocketID + 0x8
+	sizeofSocket        = sizeofSocketID + 0x18
+	SOCK_DESTROY        = 21
+)
+
+var (
+	log          = logging.DefaultLogger.WithField(logfields.LogSubsys, "datapath-sockets")
+	native       = nl.NativeEndian()
+	networkOrder = binary.BigEndian
+)
+
+type SocketDestroyer interface {
+	Destroy(filter SocketFilter) error
+}
+
+type SocketFilter struct {
+	DestIp   net.IP
+	DestPort uint16
+	Family   uint8
+	Protocol uint8
+	// Optional callback function to determine whether a filtered socket needs to be destroyed
+	DestroyCB DestroySocketCB
+}
+
+type DestroySocketCB func(id netlink.SocketID) bool
+
+// Destroy destroys sockets matching the passed filter parameters using the
+// sock_diag netlink framework.
+//
+// Supported families in the filter: syscall.AF_INET, syscall.AF_INET6
+// Supported protocols in the filter: unix.IPPROTO_UDP
+func Destroy(filter SocketFilter) error {
+	family := filter.Family
+	protocol := filter.Protocol
+
+	if family != syscall.AF_INET && family != syscall.AF_INET6 {
+		return fmt.Errorf("unsupported family for socket destroy: %d", family)
+	}
+	var errs error
+	success, failed := 0, 0
+
+	// Query sockets matching the passed filter, and then destroy the filtered
+	// sockets.
+	switch protocol {
+	case unix.IPPROTO_UDP:
```

Comment:
The current implementation only handles the UDP sockets, but I think the same issue occurs with long-live TCP connections. Why don't we support TCP as well?

Reply:
That's a fair comment.
Well, TCP has FIN/RST like signals, so if a backend is getting deleted, presumably connections are closed. At this point, we have mainly seen this issue being reported with DNS proxies using connected UDP. 
We could extend the fix for TCP connections in future if required.

Reply:
As Tim Hockin stated https://github.com/cilium/cilium/issues/23761#issuecomment-1478363344 , whether or not we should break the TCP connection comes down to what this pattern is used for. So, we can't decide to change the TCP behavior right now. It's out of the scope of this PR.